### DISCUSSION TITLE
How to Configure Rook Ceph Bucket as a Storage Class for PVC

### DISCUSSION THREAD

**Discussion by chominnn** [rook/rook#14935] [General]
## Issue: Using Rook Ceph Bucket with PersistentVolumeClaim

I am trying to use a bucket created in Rook Ceph as a storage class for a PersistentVolumeClaim, but I'm encountering issues. Below are the details of my setup.
 **Please note that some names may have been altered for privacy reasons.**

### ObjectBucketClaim

$ kubectl describe objectbucketclaims.objectbucket.io -n rook-ceph ceph--delete-bucket
Name:         ceph--delete-bucket
Namespace:    rook-ceph
Labels:       bucket-provisioner=rook-ceph.ceph.rook.io-bucket
Annotations:  <none>
API Version:  objectbucket.io/v1alpha1
Kind:         ObjectBucketClaim
Metadata:
  Creation Timestamp:  2024-10-17T05:03:42Z
  Finalizers:
    objectbucket.io/finalizer
  Generation:        5
  Resource Version:  5537176
  UID:               ba9cd6dd-c352-42a4-a59c-6aa71753da37
Spec:
  Bucket Name:           ceph-bucket-7ee8bf6b-a26e-4c13-8079-8dc7733b56d4
  Generate Bucket Name:  ceph--delete-bucket
  Object Bucket Name:    obc-rook-ceph-ceph-smart-edge-delete-bueckt
  Storage Class Name:    ceph--delete-bucket
Status:
  Phase:  Bound
Events:   <none>
### PersistentVolumeClaim

$ kubectl describe pvc -n rook-ceph ceph-pvc
Name:          ceph-pvc
Namespace:     rook-ceph
StorageClass:  ceph-sc
Status:        Pending
Volume:
Labels:        <none>
Annotations:   volume.beta.kubernetes.io/storage-provisioner: rook-ceph.ceph.rook.io/bucket
               volume.kubernetes.io/storage-provisioner: rook-ceph.ceph.rook.io/bucket
Finalizers:    [kubernetes.io/pvc-protection]
Capacity:
Access Modes:
VolumeMode:    Filesystem
Used By:       test-pod
Events:
  Type    Reason                Age                    From                         Message
  Normal  ExternalProvisioning  4m25s (x142 over 39m)  persistentvolume-controller  Waiting for a volume to be created either by the external provisioner 'rook-ceph.ceph.rook.io/bucket' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.

### StorageClass

$ kubectl describe sc ceph-sc
Name:            ceph-sc
Provisioner:     rook-ceph.ceph.rook.io/bucket
Parameters:      bucketName=ceph-bucket-7ee8bf6b-a26e-4c13-8079-8dc7733b56d,objectStoreName=rook-object-store,objectStoreNamespace=rook-ceph
AllowVolumeExpansion:  <unset>
MountOptions:          <none>
ReclaimPolicy:         Delete
VolumeBindingMode:     Immediate
Events:                <none>

### s5cmd
$ s5cmd --profile ceph --endpoint-url http://ceph-endpoint ls
2024/10/17 05:03:42  s3://ceph-bucket-7ee8bf6b-a26e-4c13-8079-8dc7733b56d

### Description
The PVC ceph-pvc is stuck in a Pending state with the message indicating that it is waiting for a volume to be created. However, the ObjectBucketClaim appears to be bound correctly.

I would appreciate any assistance or guidance on how to resolve this issue. Thank you!

**Comment by BlaineEXE** [comment]
OBCs can't be mounted at PVCs. This is a common area of confusion that I see and I think stems from the lib-bucket-provisioner's OBC project deciding to use StorageClass rather than defining a custom ObjectBucketClass resource.

The Rook documentation for OBCs explains how to use the OBC configmap and secret to consume the object store from application pods: https://rook.io/docs/rook/latest-release/Storage-Configuration/Object-Storage-RGW/ceph-object-bucket-claim/

If you want to mount S3 buckets to an application pod, you might be able to use something like AWS's mountpoint-s3: https://github.com/awslabs/mountpoint-s3/

**Reply by chominnn** [reply]
Thank you for the response.
I was trying to use the bucket as a PVC for a service like JupyterHub, where each user gets an individual PV.
It seems OBCs won’t be suitable for this purpose.
Have a great day!
